### PR TITLE
core: set up cleartext http request path

### PR DESCRIPTION
**Pull request by goaway**
Description: Adds built-in support for cleartext HTTP requests, without requiring custom config. TLS will be negotiated or not based off the request scheme. If cleartext is chosen, h2 protocol options will be ignored, and h1 will be used regardless.
Risk Level: Lowish
Testing: Local, Unit, CI

Signed-off-by: Mike Schore <mike.schore@gmail.com>

### CODE REVIEW

**Comment by carloseltuerto** [comment]
FYI, I patched the PR, but unfortunately all requests end up in the same manner:

```
EnvoyError.errorCode: 2 // likely means Unknown
EnvoyError.message: ""
EnvoyError.attemptCount: -1
EnvoyError.cause: null
```

File: library/common/http/client.cc [hunk 1]
```diff
@@ -342,46 +342,51 @@ void Client::removeStream(envoy_stream_t stream_handle) {
...
-    headers.addReference(ClusterHeader, use_h2_cluster ? BaseWlanClusterH2 : BaseWlanCluster);
-    break;
-  case ENVOY_NET_WWAN:
-    headers.addReference(ClusterHeader, use_h2_cluster ? BaseWwanClusterH2 : BaseWwanCluster);
-    break;
-  case ENVOY_NET_GENERIC:
-  default:
-    headers.addReference(ClusterHeader, use_h2_cluster ? BaseClusterH2 : BaseCluster);
-  }
+  headers.addReference(ClusterHeader, cluster);
```
Comment (carloseltuerto) [inline library/common/http/client.cc:389]: If instead this line is:

`headers.addCopy(ClusterHeader, cluster);`

Then this PR works.

Note: last time I coded in C/C++ was in previous millennium. Still, there is one thing that seems to have remained constant: using a pointer to a structure residing in the stack will end up badly if that pointer is still used once the method has returned. Editorial note: C++ is truly a counterproductive language.

Comment (goaway) [reply library/common/http/client.cc]: Yeah I missed that it was `addReference` in the initial implementation (though technically the source c string is constant, it's just the string intermediary ending up on the stack that's problematic).

**Comment by goaway** [comment]
This PR is just pending a test fix and some additional coverage that I'll have up tomorrow morning.

I have confirmed testing locally that cleartext requests appear to work, so I think the issue you're seeing now may be a separate one @carloseltuerto. But I'll be available to sync on it tomorrow.

**Comment by carloseltuerto** [comment]
> This PR is just pending a test fix and some additional coverage that I'll have up tomorrow morning.
> 
> I have confirmed testing locally that cleartext requests appear to work, so I think the issue you're seeing now may be a separate one @carloseltuerto. But I'll be available to sync on it tomorrow.

Everything is fine now - feel free to submit (as soon as you are fine with the tests you intend to update)

Thanks!

File: library/common/http/client.cc [hunk 2]
```diff
@@ -342,46 +342,56 @@ void Client::removeStream(envoy_stream_t stream_handle) {
...
-    headers.addReference(ClusterHeader, use_h2_cluster ? BaseWlanClusterH2 : BaseWlanCluster);
-    break;
-  case ENVOY_NET_WWAN:
-    headers.addReference(ClusterHeader, use_h2_cluster ? BaseWwanClusterH2 : BaseWwanCluster);
-    break;
-  case ENVOY_NET_GENERIC:
-  default:
-    headers.addReference(ClusterHeader, use_h2_cluster ? BaseClusterH2 : BaseCluster);
+  if (!h2_header.empty()) {
+    headers.remove(H2UpstreamHeader);
```
Comment (buildbreaker) [inline library/common/http/client.cc:391]: How come we didn't put this in the above if-else block instead since we have this explicit check?

Comment (goaway) [reply library/common/http/client.cc:391]: Because we need to remove it regardless of whether the first condition resolves to true.

**Review by buildbreaker — APPROVED** [review]

**Pull request merged**